[4.0] Created a HTMLHelper to render an icon

hans2103 opened the pull request:
Pull Request for Issue # .

### Summary of Changes

This PR will make it possible to use a HTMLHelper to place an icon. 
Just like there is a HTMLHelper to place an anchor and an image. 
- anchor => https://github.com/joomla/joomla-cms/blob/b2e4bf0bd26ab2bb718148459284242bcce8bc56/libraries/src/HTML/HTMLHelper.php#L253
- image => https://github.com/joomla/joomla-cms/blob/b2e4bf0bd26ab2bb718148459284242bcce8bc56/libraries/src/HTML/HTMLHelper.php#L574

### Testing Instructions

- Joomla 4 Administrator
- See the quick icons on the dashboard
- Apply patch
- See the quick icons on the dashboard ... it is still the same. Only to be implemented by the HTMLHelper now

### Actual result BEFORE applying this Pull Request

placing an icon would be: 

```
<span class="icon-joomla" aria-hidden="true"></span>
<span class="icon-joomla icon-white" aria-hidden="true"></span>
<span class="icon-joomla icon-white" aria-hidden="true" tabindex="0" title="Hello World"></span>
icon-joomla
<span class="fab fa-joomla" aria-hidden="true"></span>
```
### Expected result AFTER applying this Pull Request

```
echo HTMLHelper::_('icons.icon', 'joomla', TEXT::_('HELLO_WORLD'));
```

### Documentation Changes Required

most likely...

hans2103 committed:
Created a HTMLHelper to render an icon

hans2103 committed:
clean code

hans2103 committed:
remove hard coded icon

SharkyKZ commented:
This looks pointless and counter-intuitive.

Fedik commented on layouts/joomla/toolbar/title.php:
```diff
@@ -19,6 +21,6 @@
 ?>
 <h1 class="page-title">
-	<span class="<?php echo $icon; ?>" aria-hidden="true"></span>
+	<?php echo HTMLHelper::icon($icon); ?>
```
Correct use is `HTMLHelper::_('icon', $icon)`

hans2103 replied:
@Fedik Both ways (`HTMLHelper::_('icon'` and `HTMLHelper::icon(`) seem to work.

SharkyKZ replied:
Underscore method is the correct way. This way the method can be overridden by plugins.

hans2103 replied:
adjusted with pr https://github.com/joomla/joomla-cms/pull/30792#pullrequestreview-497366479

richard67 commented on libraries/src/HTML/HTMLHelper.php:
```diff
@@ -748,6 +748,46 @@ public static function stylesheet($file, $options = array(), $attribs = array())
+			$attribs['aria-hidden'] = 'true';
+		}
+
+		$attribs['class'] =  $iconPrefix . $icon;
```
```suggestion
		$attribs['class'] = $iconPrefix . $icon;
```
Fix PHP code style errors reported by Drone here https://ci.joomla.org/joomla/joomla-cms/35832/1/8

hans2103 committed:
https://github.com/joomla/joomla-cms/pull/30792#pullrequestreview-497366479

richard67 commented on libraries/src/HTML/HTMLHelper.php:
```diff
@@ -748,6 +748,46 @@ public static function stylesheet($file, $options = array(), $attribs = array())
+
+		if ($html !== false)
+		{
+			$attribs =  trim(ArrayHelper::toString($attribs));
```
```suggestion
			$attribs = trim(ArrayHelper::toString($attribs));
```
Fix PHP code style errors reported by Drone here https://ci.joomla.org/joomla/joomla-cms/35832/1/8

hans2103 committed:
Update libraries/src/HTML/HTMLHelper.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

hans2103 committed:
Update libraries/src/HTML/HTMLHelper.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

HLeithner commented:
Can you please remove ` $iconPrefix = 'icon-', $iconSuffix = null, $html = true, $elem = 'span',`

* We shoudn't support custom prefixes (make no sense)
* Don't know why the suffix should be an extra parameter, also php should not be used for styling
* html parameter makes no sense if you supply the complete icon "class" (without icon-)
* $elem should not be provided by php because it's a design/template decision, if I override the icon function because I want to use svg I have a problem if someone provides an unexpected markup.
* Children, don't know if this is a good idea, this would limit us by changing the iconmark to another format (which maybe can't support child elements)

dgrammatiko commented:
@hans2103 If you want to introduce a helper to render an icon then you should probably render the icon as an SVG image instead of the outdated, inefficient and mostly inaccessible font Icon. BTW there was a PR doing exactly that some time ago: https://github.com/joomla/joomla-cms/pull/28351

HLeithner commented:
@dgrammatiko if I'm write there is no way to use svg and icon fonts at the same time (override each other with css) at least I didn't find a html markup for such a case.

Using SVG sprites doesn't seem to work or only works if the svg is inline? Loading 30 svg separated doesn't seems to be the best way.

Your PR uses inline SVG sprites right and build the svg dynamically right? doesn't sound like a smart way

dgrammatiko commented:
> Loading 30 svg separated doesn't seems to be the best way.

There aren't extra Http requests, th icons are inland in the html output, read the code/comments there

HLeithner commented:
> > Loading 30 svg separated doesn't seems to be the best way.
> 
> There aren't extra Http requests, th icons are inland in the html output, read the code/comments there

That's what I mean with 'inline SVG sprites' you create a svg on the fly and reference them in another svg at the icon position.

That doesn't sound efficient at least you disable any caching

dgrammatiko commented:
> That doesn't sound efficient at least you disable any caching

Caching is useless if your rendering of a page is extremely slow but sequential loads are faster (generally speaking, for some sites this might be tolerated, eg all content is behind a login)

Anyways there are ways to do the API so you can have Icon fonts and SVG, a great example is Typo3: 
https://docs.typo3.org/m/typo3/reference-coreapi/master/en-us/ApiOverview/Icon/Index.html

hans2103 commented:
> Can you please remove ` $iconPrefix = 'icon-', $iconSuffix = null, $html = true, $elem = 'span',`
> 
> * We shoudn't support custom prefixes (make no sense)

We are working towards action named icons. An icon should be named `save`. When using this HTMLHelper it needs the icon_prefix/iconPrefix. Akeeba adds a ToolbarTitleIcon with name = akeeba. The iconPrefix is needed. 

> * Don't know why the suffix should be an extra parameter, also php should not be used for styling

Currently Joomla core has multiple implementations of extra parameters in the icon. Think about icon-spin, to let the icon-spinner icon animate and rotate. Think about icon-fw to give an icon a specific width. 

> * html parameter makes no sense if you supply the complete icon "class" (without icon-)

I agree and will remove the `$html`

> * $elem should not be provided by php because it's a design/template decision, if I override the icon function because I want to use svg I have a problem if someone provides an unexpected markup.

I agree and will remove the `$elem`

> * Children, don't know if this is a good idea, this would limit us by changing the iconmark to another format (which maybe can't support child elements)

I agree and will remove the `$children`

hans2103 committed:
remove $html and $children. Add $srOnly

hans2103 committed:
apply HTMLHelper icon including sr-only

hans2103 commented:
@dgrammatiko I prefer using svg icons too... but first things first... a HTMLHelper for the icon. Adjusting the HTMLHelper to use svg is a next step I think. The impact of this PR would be too big I think.

Looking at a11y perspective an icon should always have a text. This can be a styled label or this can be a piece of text hidden from sighted people. With the last example I mean `<span class="sr-only"></span>`.
With commit https://github.com/joomla/joomla-cms/pull/30792/commits/a3d7b64d48fca86a4d93aba9e7d68d453aa3b9f4 I have added the ability to add both icon and hidden text. Example implementation can be seen on https://github.com/joomla/joomla-cms/pull/30792/commits/3babe70869078af06e9871f3a470aa3966c30aec

brianteeman commented:
> Looking at a11y perspective an icon should always have a text.

Like all blanket statements this is not correct

hans2103 commented:
> > Looking at a11y perspective an icon should always have a text.
> 
> Like all blanket statements this is not correct

thank you for your comment. Can you rewrite it so it is correct?

dgrammatiko commented:
> but first things first... a HTMLHelper for the icon

Again IF you're about to introduce an new API endpoint make sure that 
- The arguments coulee expanded if needed (eg `icon($name = 'default', $options = [])`
- Properly deal with the static files, icons (either svg or font) are not responsibility of the template, should be handled here in this API

hans2103 commented:
> This is a remedy. A class that hides the icon for anyone using an screen reader is a hack, svg icons don't need that and thus it's the de facto best practice for 5 or so years...

Take a look at the code please. I'm not hiding the icon, I am adding a `<span class="sr-only">some text here that is not visible for sighted people, but is for screenreaders</span>`
Even svg icons need text. How often do I see that svg icons are optimized so much that the title element with a descriptive text has been removed from the svg. In our current Joomla code we accompany an icon with a sr-only span with the descriptive text. 

> The arguments could be expanded if needed (eg `icon($name = 'default', $options = []`)

@dgrammatiko am I not doing that with `$attribs = []`?

> Properly deal with the static files, icons (either svg or font) are not responsibility of the template, should be handled here in this API

can you give me an example how this should be done properly?

> Move the code in https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/HTML/Helpers/Icons.php or create a new file icon.php

that sounds good... I was not aware of the existence of this file.

dgrammatiko commented:
> @dgrammatiko am I not doing that with $attribs = []?

Read again my first comment, you **should not** spread a bunch of arguments in the function, use just 2: the icon name and then the params (an array with anything associated to with the icon you want to present). I already provided a link that explains very nicely why this is the right way...

> can you give me an example how this should be done properly?

Will do

HLeithner commented:
> I already provided a link that explains very nicely why this is the right way...

Maybe I missed it why is this the right why? for every static code analyser it's a pain because it can not check if a parameter is valid or not and since PHP 8.0 has named parameters this shouldn't be a problem at all. But maybe I'm wrong.

And why is it not the job of the template if the icon is a svg or an icon? I mean it's easy to override this icon function with your own and replace all icon fonts with svg icons (not possible with only the template you need a plugin). But I think it's not the decision of the component if the icon is a svg icon or a font.

dgrammatiko commented:
> Maybe I missed it why is this the right why? 

Argue with Uncle Bob (his book clean code is a must read for every dev): https://www.matheus.ro/2018/01/29/clean-code-avoid-many-arguments-functions/

> And why is it not the job of the template if the icon is a svg or an icon?

Simply because the icons are mostly existing in layouts and layouts should take care of their dependencies (proper modularisation). Template is essentially the orchestrator, echoing the layouts in the right order, it doesn't need to have knowledge of the specifics of every or any layout. If it does it means it is enforcing things to layouts, which is not good (a very good example is https://github.com/joomla/joomla-cms/pull/23742 )

HLeithner commented:
> Argue with Uncle Bob (his book clean code is a must read for every dev): https://www.matheus.ro/2018/01/29/clean-code-avoid-many-arguments-functions/

So my conclusion from this text is, that I did it right ;-) I requested to reduce the parameter count (see $html and so on). But it seems you missed the important part, if i'm not wrong he uses Java as an example in refactoring "name, phone, street..." to an Object. Your suggest was to add it to an assoc array that's not the same and error prone. The Object he created is "Address" which likely knows exactly at compile time which parameters can be added. We can do the same by creating an class "iconObject" which hold the needed information so the runtime or a static code analyzer can print an error if you have a typo. Which is not so easy to do with an assoc array.

And I think for this utility functions is a bigger pain to have an object instead of 2 additional parameters (which I would remove anyway, still don't understand why we have a suffix)

> Simply because the icons are mostly existing in layouts and layouts should take care of their dependencies (proper modularisation). Template is essentially the orchestrator, echoing the layouts in the right order, it doesn't need to have knowledge of the specifics of every or any layout. If it does it means it is enforcing things to layouts, which is not good (a very good example is #23742 )

so what's the suggestion on this? I have the "idea" or some thoughts that if this PR is finished and the icon pr is finished (and if it will be merged) we should be possible to simply switch all (at least core icons) simple to svgs (which a variant of your last pr).

dgrammatiko commented:
>  we should be possible to simply switch all (at least core icons) simple to svgs 

That would be the perfect scenario, if we can agree on that and write some code to get us there 😉

hans2103 commented:
@dgrammatiko and @HLeithner I like both of your suggestions and I look forward to the moment we can use svg instead of fonts. 
Can you make a PR on my PR with your suggestions? Doing so we can all work towards a nice solution.
https://github.com/hans2103/joomla-cms/tree/htmlhelper-for-an-icon

HLeithner commented:
@hans2103 small steps as you say ;-) and please move the pr to the icons htmlhelper class.

I think dimitris suggest in his edit and original PR is a bit to over-engineered, I think we don't need to add the provider i'm thinking the other way around (mainly for core), we as core provide x icon functions. If you want to use them you can use it, if you need another one you have to do your own inclusion/registration what ever. Also setting the "type" (svg/font) at the layout level is wrong and you can't override it in a template.

And if all our icons have to use such a method we hit the most complexity possible.

```php
<?php echo HTMLHelper::_('icon.icon', 'info', 'svg', ['provider' => 'fontawesome-free', 'group' => 'solid', 'classes' => ['fa', 'demo-class'], 'attributes' => ['data-foo' => 'bar', 'id' => 'foobar']], 'sr-text' => Text::_('INFO')]); ?>
```
I mean nobody will use this.

dgrammatiko commented:
@HLeithner actually if someone is using Fontawesome (due to the defaults) the code will be way simpler:

```php
<?php echo HTMLHelper::_('icon.icon', 'info', 'svg', [ 'classes' => ['fa', 'demo-class'], ''sr-text' => Text::_('INFO')]); ?>
```

hans2103 commented:
The expected output is 
```
<span class="fa fas-info"></span>
<span class="sr-only"><?php echo Text::_('INFO')]); ?></span>
```

the already existing styling for sr-only is
```
.sr-only {
    border: 0;
    clip: rect(0,0,0,0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
    white-space: nowrap;
}
```

This piece of styling reduces the element to a non-visible but still existing element. 
Combining `sr-only` with the icon will result in a non-visible icon. 

Dinner now... will check on your suggestions later today or tomorrow morning. Thank you for your effort.

dgrammatiko commented:
> Combining sr-only with the icon will result in a non-visible icon.

Then the code becomes even simpler:

```php
<?php echo HTMLHelper::_('icon.icon', 'info', 'font'); ?>
// 2 Arguments: Name of the icon, expected format
```

hans2103 commented:
as a frontend developer I stumble upon my lack of decent PHP knowledge to create the content of the new /icon.php file. 
@dgrammatiko and / or @HLeithner can you create a PR on this PR with your suggested /Icon.php file?

Fedik commented:
just put your `function icon(...) ...` code in to this one https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/HTML/Helpers/Icons.php
and then can call as 
```
echo HTMLHelper::_('icons.icon', ....
```

dgrammatiko commented:
> can you create a PR on this PR with your suggested /Icon.php file

I need to do some research on the FA css/fonts, probably this weekend

hans2103 committed:
move function to libraries/src/HTML/Helpers/Icons.php

hans2103 commented:
as suggested.. I've moved the function to libraries/src/HTML/Helpers/Icons.php
and adjusted the call with `HTMLHelper::_('icons.icon'....`

hans2103 commented:
if the output is:
```
<span class="fa fas-info"></span>
<span class="sr-only"><?php echo Text::_('INFO')]); ?></span>
```

and the input is
```
<?php echo HTMLHelper::_('icons.icon', 'info', 'font', [  'sr-text' => Text::_('INFO'), ]); ?>
```


should we add a JLayout too where icon-prefix, extra classNames are handled?

dgrammatiko commented:
> should we add a JLayout too where icon-prefix, extra classNames are handled?

The second argument can have prefix suffix or am I getting something wrong (eg `foo-icon` or `icon-bar`)

hans2103 commented:
> > should we add a JLayout too where icon-prefix, extra classNames are handled?
> 
> The second argument can have prefix suffix or am I getting something wrong (eg `foo-icon` or `icon-bar`)

What do you think about not handling the icon-prefix over here in the HTMLHelper, but set it using the JLayout?
That way this HTMLHelper stays a functional thing forwarding some stuff and let the JLayout handle the HTML .

degobbis commented:
Here my 2 cents, if relevant.
I also think the parameters `$iconPrefix` and `$iconSuffix` are not needed.

Let's take the change from `layouts/joomla/searchtools/default/noitems.php`:
The expected output for the icon is `<span class="fas fa-info-circle" aria-hidden="true"></span>`, but if we use your code now, we get `<span class="icon-icon-fas fa-info-circle" aria-hidden="true"></span>`.

This results because the prefix is defined by default and also set for a second time in `$attribs['class'] = $iconPrefix . $icon;`.

So if the prefix is not set by default, a developer has to consider this individually, then he can write it directly into the actual class and the parameters become unneeded. The same applies to the suffix.

hans2103 committed:
:wastebasket: housecleaning and apply suggestions by @degobbis

hans2103 committed:
add missing className

hans2103 commented:
Updated implementation:

```
<?php echo HTMLHelper::_('icons.icon', 'fas fa-info-circle', Text::_('INFO')); ?>
```

after merge of #30707 this has to change to:
```
<?php echo HTMLHelper::_('icons.icon', 'icon-info-circle', Text::_('INFO')); ?>
```

Method for implementation has been simplified to:
(asa @dgrammatiko suggested in https://github.com/joomla/joomla-cms/pull/30792#issuecomment-702028859)
```
	/**
	 * Method to write a `<span>` element for an icon
	 *
	 * @param   string        $icon        The funcitonal name for an icon.
	 * @param   string        $srOnly      Screen Reader text if no visible text is placed
	 * @param   array         $attribs     Attributes to be added to the wrapping element
	 *
	 *
	 * @return  string
	 *
	 * @since   4.0
	 */
	public static function icon($icon = 'default', $srOnly = null, $attribs = [])
	{
		if (isset($attribs['class']))
		{
			$icon = $icon . ' ' . $attribs['class'];
		}

		$attribs['class'] = $icon;

		if (!isset($attribs['aria-hidden']))
		{
			$attribs['aria-hidden'] = 'true';
		}

		if (isset($srOnly))
		{
			$text = htmlspecialchars($srOnly, ENT_COMPAT, 'UTF-8');
			$srOnly = '<span class="sr-only">' . $text . '</span>';
		}

		return '<span ' . trim(ArrayHelper::toString($attribs)) . '></span>' . $srOnly;
	}
```

hans2103 committed:
Merge branch '4.0-dev' into feature/htmlhelper-for-an-icon

Quy commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,41 @@
+	/**
+	 * Method to write a `<span>` element for an icon
+	 *
+	 * @param   string        $icon        The funcitonal name for an icon.
```
Per Joomla! Coding Standards:

> Commonly a line after the tag @param consists of the following three parts in order of appearance:
> 
> * variable type (There must be 3 spaces before variable type.)
> * variable name (There must be 2 spaces after the longest type.)
> * variable description (There must be 2 spaces after the longest variable name.)
> 
> If there are more than one @param the type, names and description have to be aligned.

Typo `funcitonal`

hans2103 replied:
fixed suggestions in commit https://github.com/joomla/joomla-cms/pull/30792/commits/1a950bfabc28767c52d00b1b7d1bb9bf10a35bcc

Quy commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,41 @@
+	 * @param   string        $icon        The funcitonal name for an icon.
+	 * @param   string        $srOnly      Screen Reader text if no visible text is placed
+	 * @param   array         $attribs     Attributes to be added to the wrapping element
+	 *
+	 *
+	 * @return  string
+	 *
+	 * @since   4.0
```
```suggestion
	 *
	 * @return  string
	 *
	 * @since   __DEPLOY_VERSION__
```

Quy commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,41 @@
+	{
+		if (isset($attribs['class']))
+		{
+			$icon = $icon . ' ' . $attribs['class'];
```
```suggestion
			$icon .= ' ' . $attribs['class'];
```

hans2103 committed:
Merge branch 'upstream/4.0-dev' into feature/htmlhelper-for-an-icon

hans2103 committed:
Update libraries/src/HTML/Helpers/Icons.php

Co-authored-by: Quy <quy@fluxbb.org>

hans2103 committed:
Update libraries/src/HTML/Helpers/Icons.php

Co-authored-by: Quy <quy@fluxbb.org>

hans2103 committed:
apply Joomla! Coding Standards for DocBlock

hans2103 committed:
replace null by '', since variable is a string and default value null is not

hans2103 committed:
check if variable isset and not empty

Quy commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,40 @@
+			$attribs['aria-hidden'] = 'true';
+		}
+
+		if (isset($srOnly) && !empty($srOnly))
```
```suggestion
		if (!empty($srOnly))
```

hans2103 committed:
Update libraries/src/HTML/Helpers/Icons.php

Co-authored-by: Quy <quy@fluxbb.org>

SharkyKZ commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,40 @@
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function icon($icon = 'default', $srOnly = '', $attribs = [])
```
```suggestion
	public static function icon(string $icon = 'default', string $srOnly = '', array $attribs = [])
```

SharkyKZ commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,40 @@
+			$attribs['aria-hidden'] = 'true';
+		}
+
+		if (!empty($srOnly))
```
```suggestion
		if ($srOnly !== '')
```

hans2103 committed:
Update libraries/src/HTML/Helpers/Icons.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

hans2103 committed:
Update libraries/src/HTML/Helpers/Icons.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

Quy commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,40 @@
+			$srOnly = '<span class="sr-only">' . $text . '</span>';
+		}
+
+		return '<span ' . trim(ArrayHelper::toString($attribs)) . '></span>' . $srOnly;
```
Is is necessary to apply `trim`?

hans2103 replied:
@quy I think I've added it because I saw obsolete spacing at the end.

Quy replied:
Here is a var_dump on `ArrayHelper::toString($attribs)` without `trim` and the result:

> string(63) "class="fas fa-fas fa-home cpanel icon-white" aria-hidden="true""

hans2103 replied:
removed in commit a414b8467f8d47dffe4992316154510b04467ecd

hans2103 committed:
removal trim

hans2103 committed:
#30707 has not been merged yet. :-)

Quy commented:
I have tested this item :white_check_mark: successfully on 846466f5bc47c80e876825cc9b82b968010abb74<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30792">issues.joomla.org/tracker/joomla-cms/30792</a>.</sub>

SharkyKZ commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,40 @@
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function icon(string $icon = 'fas fa-default', string $srOnly = '', array $attribs = [])
```
There is no `fas fa-default` icon. Why even add a default here?

hans2103 replied:
adding a default saves me from adding a check if the icon is set or not. 
I will add the check and return null if no icon is set.

hans2103 replied:
see commit 62c833c2ddd83d17bb8582fcf2cc33726b1a3aed

hans2103 committed:
remove default icon and add if-then statement to check existence of icon

SharkyKZ commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,45 @@
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function icon(string $icon = '', string $srOnly = '', array $attribs = [])
```
Icon should not be optional.

```suggestion
	public static function icon(string $icon, string $srOnly = '', array $attribs = [])
```

SharkyKZ commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +22,45 @@
+	{
+		if ($icon === '')
+		{
+			return '';
```
Maybe overkill but do you want to throw an exception here?

hans2103 replied:
@SharkyKZ I never implemented an exception. Can you create a PR on mine, so I can learn from it?

SharkyKZ replied:
Change return to `throw new \InvalidArgumentException('error message here'); and document with `@throws`.

Again, I'm not sure if this is not overkill. I see only one instance of exception being thrown in HTML helpers. But maybe that's because we have poor sanity checks in general.

hans2103 replied:
@SharkyKZ implemented with commit https://github.com/joomla/joomla-cms/pull/30792/commits/04c3854357a557020e9d7bf5e11a1a4ddb1ba862

hans2103 committed:
Merge branch 'upstream/4.0-dev' into feature/htmlhelper-for-an-icon

hans2103 committed:
replace return '' by throw

hans2103 committed:
Update libraries/src/HTML/Helpers/Icons.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

Quy commented on administrator/language/en-GB/lib_joomla.ini:
```diff
@@ -502,7 +502,8 @@ JLIB_HTML_FEATURED_ITEM="Featured."
 JLIB_HTML_GOTO_PAGE="Go to page %s"
-JLIB_HTML_GOTO_POSITION="Go to %s page"
+JLIB_HTML_GOTO_POSITION="Go to %s page."
+JLIB_HTML_ICONS_NO_ICON="No icon provided
```
```suggestion
JLIB_HTML_ICONS_NO_ICON="No icon provided"
```

Quy replied:
```suggestion
JLIB_HTML_ICONS_NO_ICON="No icon provided."
```

SharkyKZ commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -21,6 +23,47 @@
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function icon(string $icon, string $srOnly = '', array $attribs = [])
```
```suggestion
	public static function icon(string $icon, string $srOnly = '', array $attribs = []): string
```

hans2103 committed:
Update administrator/language/en-GB/lib_joomla.ini

Co-authored-by: Quy <quy@fluxbb.org>

hans2103 committed:
Update libraries/src/HTML/Helpers/Icons.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

N6REJ commented:
@hans2103 you might find https://github.com/joomla/joomla-cms/pull/31190 interesting.

hans2103 committed:
Merge branch 'upstream/4.0-dev' into feature/htmlhelper-for-an-icon

hans2103 commented:
> @hans2103 you might find #31190 interesting.

Can your svg improvement be merged with this PR?

Quy commented:
I have tested this item :white_check_mark: successfully on b1660ab4e5e736a123061bccd9df953dc79ffc50<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30792">issues.joomla.org/tracker/joomla-cms/30792</a>.</sub>

N6REJ commented:
> > @hans2103 you might find #31190 interesting.
> 
> Can your svg improvement be merged with this PR?

baby steps.. mine seems to be up for debate as always.

infograf768 commented:
Please, when modifying the lib_joomla.ini file, modify both files (admin and site).
Can't be merged until done.

hans2103 committed:
added translation to frontend too

hans2103 commented:
@infograf768 just added the language string to the fronted with commit https://github.com/joomla/joomla-cms/pull/30792/commits/631ab068f081f727295aa4dc1bedbb8dd2a26a35

infograf768 commented on language/en-GB/lib_joomla.ini:
```diff
@@ -503,6 +503,7 @@ JLIB_HTML_FEATURED_PENDING_ITEM="Featured, but is Pending."
 JLIB_HTML_FEATURED_STARTED="Start: %s"
 JLIB_HTML_GOTO_PAGE="Go to page %s"
 JLIB_HTML_GOTO_POSITION="Go to %s page"
```
Missing period

hans2103 replied:
removed period in admin lib_joomla.ini with commit cb05e071b6f364dc7f3f5241ec0084fef2ee8dad
Is out of scope for this PR.

hans2103 committed:
removal period

hans2103 committed:
Merge branch 'upstream/4.0-dev' into feature/htmlhelper-for-an-icon

hans2103 committed:
Merge branch 'upstream/4.0-dev' into feature/htmlhelper-for-an-icon

# Conflicts:
#	layouts/joomla/searchtools/default/noitems.php
#	layouts/joomla/toolbar/title.php

hans2103 committed:
reapply change

Quy commented:
I have tested this item :white_check_mark: successfully on b1660ab4e5e736a123061bccd9df953dc79ffc50<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30792">issues.joomla.org/tracker/joomla-cms/30792</a>.</sub>

N6REJ commented:
I have tested this item :white_check_mark: successfully on b1660ab4e5e736a123061bccd9df953dc79ffc50<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30792">issues.joomla.org/tracker/joomla-cms/30792</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30792">issues.joomla.org/tracker/joomla-cms/30792</a>.</sub>

dgrammatiko commented:
@quy please remove Rtc from this there are a number of issues I pointed in my comments that are not addressed yet

hans2103 commented:
@dgrammatiko please describe them again. 
Or can they be fixed in a new PR?

dgrammatiko commented:
It's really bad idea to merge a half done API so it cannot be done in a new pr, needs to be fixed here. Read my comment about handling the related assets and also the api needs to be able to either echo font icon or svg icon html

hans2103 commented:
@dgrammatiko 
do you mean this part of the comment copied from https://github.com/joomla/joomla-cms/pull/30792#issuecomment-700006784
> - Since you're introducing an API for icons it seems that you're missing the most important aspect: decouple the assets. Right now template are dictating the behaviour (eg loading the font always) which -A. it's not template'e responsibility -B. the API should load any assets required. Eg. If my landing page doesn't have any icons why does the template should enforce the loading of unneeded css and fonts? (the template has no clue what each layout is using thus the only safe place to add assets in in the layouts NOT IN the template as these are globally loaded [monolithic approach, really outdated])

hans2103 committed:
Merge branch 'upstream/4.0-dev' into htmlhelper-for-an-icon

# Conflicts:
#	libraries/src/HTML/Helpers/Icons.php

hans2103 committed:
FontAwesome provides svg icons... use them

hans2103 committed:
no idea how to explain the test instructions for this. :-)

hans2103 committed:
easier for test instructions... after apply path the icons on dashboard should be visible

if helper is finished

hans2103 committed:
add new parameter $type to trigger svg icons

dgrammatiko commented on build/build-modules-js/settings.json:
```diff
@@ -255,6 +255,7 @@
         },
         "filesExtra": {
           "scss": "scss",
+          "svgs": "svgs",
```
This needs to be`"svgs": "images",`

hans2103 replied:
solved in commit 83551d9a7076d514fb4c193773c11f2cb94c74e0

hans2103 commented:
Adjusted `icons.icon` to get triggered by `type = svg`
Added the FontAwesome SVG icons

But... help needed to call for the FontAwesome SVG icons. 

Should a reference to /media/vendor/fortawesome-free/svgs/regular be made within the function icon? => I hope not... it is not up to this function to know where the svg icons are. 
Or should this be done via a JLayout where an easy override can be chosen for another library? => makes sense to me.

dgrammatiko commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -169,4 +130,52 @@ public static function svg(string $file, bool $relative = true): ?string
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function icon(string $icon, string $srOnly = '', string $type = '', array $attribs = []): string
```
Please remove the srOnly or move it as a param to attribute. Actually please bring the logic from my PR https://github.com/joomla/joomla-cms/pull/28351/files#diff-1f3d4b0ff411d275d94c1538ef487ef75d4450d772ad2cf712b5796467917f6dR214-R246
That allows any type of attributes to be attached to the inserted svg...

hans2103 replied:
applied your changes... some of them. 
I don't want to use svg-sprite at this moment... out of scope for this change. 

But.. I need some help to get provide the correct path to the function `icons.svg`

suggestions?

dgrammatiko replied:
> I don't want to use svg-sprite at this moment... out of scope for this change.

You're missing the point my friend, if you're not controlling the output (eg current implementation) you're essentially allowing broken HTML to be applied in any page. Let me explain: the svg helper will echo each fragment every time you call it, meaning it can inflate the output but worse than that it could echo multiple svgs with the same id which is invalid HTML. Unfortunately if you want to do SVGs correctly you need somehow to take control of the fragments you're inserting or the API you're introducing is prone to undesired side effects

hans2103 committed:
adjust settings.json

hans2103 committed:
call helper icons.icon with svg

hans2103 committed:
optimize icons.icon for svg

SharkyKZ commented:
This PR makes even less sense now.

brianteeman commented on libraries/src/HTML/Helpers/Icons.php:
```diff
@@ -128,4 +130,57 @@ public static function svg(string $file, bool $relative = true): ?string
+
+		$attribs['class'] = $class;
+
+		if (!isset($attribs['aria-hidden']))
+		{
+			$attribs['aria-hidden'] = 'true';
+		}
+
+		if (isset($attribs['text']))
+		{
+			$text = htmlspecialchars($attribs['text'], ENT_COMPAT, 'UTF-8');
+			$text = '<span class="sr-only">' . $text . '</span>';
+			unset($attribs['text']);
+		}
```
It wouldnt make sense to have sr-only text if for some reason aria-hidden is not true

hans2103 replied:
Attribute `aria-hidden` is always true, unless it is set otherwise. 
Do you mean that if it is set otherwise, the span element with className `sr-only` should not be set?

brianteeman replied:
yes

hans2103 replied:
solved in commit 4a765c547bdf4761538a8ab3b8f0044f4ce616d2

hans2103 commented:
@SharkyKZ 
> This PR makes even less sense now.

could you explain me why?

hans2103 committed:
only set sr-only text when aria-hidden is true
